### PR TITLE
Combat overhaul - Armor calculations

### PR DESCRIPTION
**Pull request by Gandalf2k15**
## About The Pull Request

Armor degrades over time if its armor type is hit, this is a percentage calculation. All armor starts at 100%.

Armor only degrades if the incoming damage type is something the armor is rated for.

Armor can be fixed with armor plates.

Bullet damage is calculated by checking

Armor damage rating is calculated by the base armor value then minus its integrity percent. E.g. 89% of 30
If bullet damage is more than the armor rating, the armor rating delta of the bullet damage is converted into stamina, the excess damage is dealt as the bullet damage type.
Some bullets have an armour_penetration value assigned, this is taken away from the armor rating.

Calculation for damage:
B = standard bullet damage
I = Armor integrity
AP = Bullet armor penetration
R = Armor rating
D = Dealt damage
S = Dealt stamina damage

D = B - ((R / 100 * I) - AP)
S = B - D

Calculation for integrity loss:
Bullet damage / 10

If a bullet has a higher damage than armor rating, it will penetrate and cause a wound, while also dealing the leftover delta damage.

Example: a 45 brute damage bullet hitting 30 brute armor would result in 30 stamina damage and 15 brute. Armor would degrade by 4.5%, 45 / 10. Next 45 damage bullet would deal 28.65 stamina and 16.35 brute.

After the pain system is introduced, the stamina damage dealt will be split between pain and stamina.

:cl:
balance: Armor is now far more relevant, see: https://github.com/Skyrat-SS13/Skyrat-tg/pull/12086
/:cl:

### CODE REVIEW

File: modular_skyrat/modules/combat_overhaul/code/armor.dm [hunk 1]
```diff
@@ -0,0 +1,15 @@
+/datum/armor
+	/// Our integrity, from 0 - 100 percent.
+	var/integrity = 100
+
+/datum/armor/getRating(rating, ignore_integrity) // THIS IS A PROC OVERRIDE
+	if(ignore_integrity)
```
Comment (Zonespace27) [inline modular_skyrat/modules/combat_overhaul/code/armor.dm]: could probably be a one-line ternery

File: modular_skyrat/modules/combat_overhaul/code/clothing.dm [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#define REQUIRED_REPAIR_STACK 5
```
Comment (Zonespace27) [inline modular_skyrat/modules/combat_overhaul/code/clothing.dm:1]: undef these

File: modular_skyrat/modules/combat_overhaul/code/clothing.dm [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+/obj/item/clothing/examine(mob/user)
+	. = ..()
+	if(armor.bio || armor.bomb || armor.bullet || armor.energy || armor.laser || armor.melee || armor.fire || armor.acid)
+		. += "The armor plating is at [armor.integrity]% integrity."
+
+/obj/item/clothing/attackby(obj/item/attacking_item, mob/user, params)
+	if(istype(attacking_item, /obj/item/stack/sheet/iron) && (armor.bio || armor.bomb || armor.bullet || armor.energy || armor.laser || armor.melee || armor.fire || armor.acid))
+		var/obj/item/stack/sheet/iron/iron_stack = attacking_item
+		if(!iron_stack.use(REQUIRED_REPAIR_STACK))
+			to_chat(user, span_danger("Not enough metal to repair the armor plating!"))
```
Comment (Zonespace27) [inline modular_skyrat/modules/combat_overhaul/code/clothing.dm]:
```suggestion
			to_chat(user, span_danger("You don't have enough of [attacking_item] to repair [src]'s armor plating!"))
```

File: modular_skyrat/modules/combat_overhaul/code/clothing.dm [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+		if(do_after(user, REPAIR_TIME, src))
```
Comment (Zonespace27) [inline modular_skyrat/modules/combat_overhaul/code/clothing.dm:22]: early return

File: modular_skyrat/modules/combat_overhaul/code/clothing.dm [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+			to_chat(user, span_notice("You repair the armor plating."))
```
Comment (Zonespace27) [inline modular_skyrat/modules/combat_overhaul/code/clothing.dm]:
```suggestion
			to_chat(user, span_notice("You repair [src]'s armor plating."))
```

**Review by Zonespace27 — CHANGES_REQUESTED** [review]

**Comment by YakumoChen** [comment]
I'm not sure about this one chief. There's already a big overhaul for armour coming down the pipeline. Does this change armour values? 

This really kind of sounds like it's going to either make armour functionally useless (you'll get wrecked by stamloss even if you reduce most projectile damage) or it's going to be a colossal buff for security (which naturally has better access to armour and is migitating a lot of damage, than most non-loud antags)

Armour degradation is a bigger problem for antags than it is for security regardless (who have advantage via numbers, guaranteed lathe access), so I'm pretty iffy about it.

**Comment by Gandalf2k15** [comment]
> I'm not sure about this one chief. There's already a big overhaul for armour coming down the pipeline. Does this change armour values?
> 
> This really kind of sounds like it's going to either make armour functionally useless (you'll get wrecked by stamloss even if you reduce most projectile damage) or it's going to be a colossal buff for security (which naturally has better access to armour and is migitating a lot of damage, than most non-loud antags)
> 
> Armour degradation is a bigger problem for antags than it is for security regardless (who have advantage via numbers, guaranteed lathe access), so I'm pretty iffy about it.

It changes no armor values and will work nicely with the update, it's also a massive buff all round but also a massive debuff because antags can easily disable an officer. And armor integrity can be repaired by metal. That's very abundant and anyone has access to the orm.

**Comment by Paxilmaniac** [comment]
every day we stray closer to warzone, it starts with the armor plates the future is doomed

**Comment by YakumoChen** [comment]
>  That's very abundant and anyone has access to the orm.

ORM requires a specific access that some jobs (assistants) do not have, actually.

**Comment by Gandalf2k15** [comment]
https://github.com/tgstation/tgstation/pull/65053 is a better option

**Pull request closed without merging**